### PR TITLE
fix: show validation errors when going back to a page

### PR DESCRIPTION
**Pull request by jamil314**
## Description

- Mark all the fields of a form page as touched as soon as we leave the page.
-  Refactor `Name` field to be more granular. This allows us to have error messages on each part of the name field (e.g: 'firstname', 'surname') individually.

## Checklist

- [x] I have linked the correct Github issue under "Development"
- [x] I have tested the changes locally, and written appropriate tests
- [x] I have tested beyond the happy path (e.g. edge cases, failure paths)
- [ ] I have updated the changelog with this change (if applicable)
- [x] I have updated the GitHub issue status accordingly

### CODE REVIEW

File: packages/client/src/v2-events/components/forms/FormFieldGenerator/FormFieldGenerator.tsx [hunk 1]
```diff
@@ -120,14 +126,31 @@ export const FormFieldGenerator: React.FC<FormFieldGeneratorProps> = React.memo(
              * have been touched for once during the form manipulation. So that we can show validation
              * errors for all fields that have been touched.
              */
-            if (
+            const newlyTouched =
               Object.keys(touched).length > 0 &&
               !isEqual(touched, initialTouchedFields) &&
-              Object.keys(touched).some((key) => !(key in initialTouchedFields))
-            ) {
+              Object.keys(touched).filter(
```
Comment (makelicious) [inline packages/client/src/v2-events/components/forms/FormFieldGenerator/FormFieldGenerator.tsx]: IT starting to get quite complex. Could this be a function?

File: packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx [hunk 2]
```diff
@@ -327,7 +331,18 @@ export function FormSectionComponent({
                   fieldDefinition={field}
                   form={completeForm}
                   readonlyMode={readonlyMode}
-                  touched={touched[field.id] ?? false}
+                  touched={
+                    /**
```
Comment (makelicious) [inline packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx:335]: Will this work for address, too?

File: packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx
Comment (makelicious) [inline packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx:1]: We could generalize the error mssaging. name should just be an example. Whatever we do, should preferably work on all nested field types.

File: packages/commons/src/events/FieldConfig.ts [hunk 3]
```diff
@@ -322,26 +322,41 @@ export const SelectDateRangeField = BaseField.extend({
...
+
+export type NameConfig = z.infer<typeof NameConfig>
+
 const NameField = BaseField.extend({
   type: z.literal(FieldType.NAME),
   defaultValue: z
     .object({
-      firstname: NonEmptyTextValue,
-      surname: NonEmptyTextValue
+      firstname: NonEmptyTextValue.optional(),
```
Comment (makelicious) [inline packages/commons/src/events/FieldConfig.ts:337]: I like this better than the includeMiddlename flag. However, we need to make sure it is excluded in the payload.

File: packages/commons/src/events/FieldConfig.ts [hunk 4]
```diff
@@ -322,26 +322,41 @@ export const SelectDateRangeField = BaseField.extend({
...
+      middlename: NonEmptyTextValue.optional(),
+      surname: NonEmptyTextValue.optional()
     })
     .optional(),
   configuration: z
     .object({
+      nameConfig: NameConfig.default({
```
Comment (makelicious) [inline packages/commons/src/events/FieldConfig.ts]: I'm thinking how could we make this more unified with all the other configurations. We can introduce this new pattern but it is about being required.


option 1:

Complex field would allow two types of config.

`required?: boolean | { firstname: boolean; surname: boolean; middlename: boolean; }` 

option 2:
configuration object would be recursive:

```
{
   configuration: {
      maxLength: z.number().optional().describe('Maximum length of the text'),
      prefix: TranslationConfig.optional(),
      postfix: TranslationConfig.optional(),
      firstname: {
        required: true,
      }
      surname: {
        required: false
      }
      // or wrapped to the way you have done
      name: { // or fields
         firstname: { required: true },
         surname: { required: true }
      }
      
   }
 }
```
What do you think?

**Review by makelicious — COMMENTED** [review]
Nice idea! We need test cases for the name component, and for the validation error fix you performed.

Left suggestions on the approach.

File: packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx (hunk 2 above)
Comment (jamil314) [reply packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx:335]: It will

File: packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx
Comment (jamil314) [reply packages/client/src/v2-events/components/forms/FormFieldGenerator/FormSectionComponent.tsx:1]: We just need to pass the parentId to `FormFieldGenerator`, and it will work.

**Review by rikukissa — APPROVED** [review]
